tests/lib/mem_alloc/src/main.c

```diff
@@ -0,0 +1,237 @@
+	char *reloc_ptr = NULL;
+
+	errno = NO_ERROR;
+
+	ptr = (char *) malloc(orig_size);
+
+	zassert_not_null((ptr), "malloc failed, errno: %d", errno);
+	zassert_equal(errno, 0, NULL);
+	memset(ptr, 'p', orig_size);
+
+	reloc_ptr = (char *) realloc(ptr, new_size);
+
+	zassert_not_null(reloc_ptr, "realloc failed, errno: %d", errno);
+	zassert_not_null((ptr), "malloc/realloc failed, errno: %d", errno);
+	zassert_equal(errno, 0, NULL);
+	ptr = reloc_ptr;
+
+	memset(filled_buf, 'p', BUFSIZE);
+	zassert_true(((memcmp((char *)ptr, filled_buf, BUFSIZE)) == 0),
+			"realloc failed to copy malloc data, errno: %d", errno);
+
+	free(ptr);
+	ptr = NULL;
+}
+
+/**
+ * @brief Test dynamic memory allocation using reallocarray
+ *
+ * @see malloc(), reallocarray(), free()
+ */
+#ifdef CONFIG_NEWLIB_LIBC
+void test_reallocarray(void)
+{
+	/* reallocarray not implemented for newlib */
+	ztest_test_skip();
+}
+#else
+void test_reallocarray(void)
+{
+	char orig_size = BUFSIZE;
+	char *ptr = NULL;
+
+	errno = NO_ERROR;
+	ptr = (char *) malloc(orig_size);
+
+	zassert_not_null((ptr), "malloc failed, errno: %d", errno);
+	zassert_equal(errno, 0, NULL);
+	memset(ptr, 'p', orig_size);
+
+	char *reloc_ptr = (char *) reallocarray(ptr, 2, orig_size);
+
+	zassert_not_null(reloc_ptr, "reallocarray failed");
+	zassert_not_null((ptr), "malloc/reallocarray failed, errno: %d", errno);
+	zassert_equal(errno, 0, NULL);
+	ptr = reloc_ptr;
+
+	memset(filled_buf, 'p', BUFSIZE);
+	zassert_true(((memcmp((char *)ptr, filled_buf, BUFSIZE)) == 0),
+			"realloc failed to copy malloc data, errno: %d", errno);
+
+	free(ptr);
+	ptr = NULL;
+}
+#endif
+
+
+#define MAX_LEN (10 * BUF_LEN)
+
+/**
+ * @brief Test dynamic memory allocation functions
+ *
+ * @see malloc(), calloc(), realloc(), free()
+ */
+void test_memalloc_all(void)
+{
+	char *mlc_ptr = NULL;
+	char *clc_ptr = NULL;
+	char *reloc_ptr = NULL;
+	int orig_size = BUF_LEN;
+	int new_size = MAX_LEN;
+
+	mlc_ptr = (char *) malloc(orig_size);
+	zassert_not_null((mlc_ptr), "malloc failed, errno: %d", errno);
+
+	clc_ptr = (char *) calloc(100, sizeof(char));
+	zassert_not_null((clc_ptr), "calloc failed, errno: %d", errno);
+
+	reloc_ptr = (char *) realloc(mlc_ptr, new_size);
+	zassert_not_null(reloc_ptr, "realloc failed, errno: %d", errno);
+	zassert_not_null((mlc_ptr), "malloc/realloc failed, errno: %d", errno);
+	mlc_ptr = reloc_ptr;
+
+	free(mlc_ptr);
+	free(clc_ptr);
+	mlc_ptr = NULL;
+	clc_ptr = NULL;
+	reloc_ptr = NULL;
+}
+
+/**
+ *
+ * @brief Test dynamic memory allocation upto maximum size
+ * Negative test case
+ *
+ */
+
+void test_memalloc_max(void)
+{
+	size_t max_heap_size = 0;
+	char *ptr = NULL;
+
+#ifdef CONFIG_NEWLIB_LIBC
+	void *heap_base = NULL;
+
+	z_newlib_get_heap_bounds(&heap_base, &max_heap_size);
+#else
+	max_heap_size = MAX_HEAP_SIZE;
+#endif
+	ptr = (char *) malloc(max_heap_size + 1);
+	zassert_is_null(ptr, "malloc passed unexpectedly");
+	zassert_equal(errno, ENOMEM, "Wrong error number, errno: %d", errno);
```

Comment:
remove this check.
malloc() setting errno is in the C99 spec, but not in C11 spec, and newlib currenty is not setting errno in this situation.

Reply:
Sure, i did refer minimal_libc malloc implementation that sets errno.
Removed to allow test pass for both minimal and newlib implementation.
Thanks for the information.